### PR TITLE
live timers: API boundary, implementation, opt-in

### PR DESCRIPTION
**Pull request by yaauie**
## Release notes

 [rn:skip]

## What does this PR do?

Introduces an API boundary for timers as a first-class metric, as described in elastic/logstash#14675, and migrates all known internal timers to use the new API boundary for tracked execution.

Please refer to the specification for details on motivations.

## Why is it important/What is the impact to the user?

STUB: as of PR creation, there is no impact to the user. Developers using the Ruby- or Java-based Plugin APIs will have access to a new API boundary for acquiring a timer object that they can use to track execution time.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~[ ] I have made corresponding changes to the documentation~ _[deferred: #14809]_
- [x] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] introduce write-only `TimerMetric` to  Java API
- [x] introduce read-write internal `TimerMetric` interface that extends `Metric<Long>`
- [x] implement our internal `TimerMetric` with _status-quo_ `AfterCompletionTimerMetric`
- [x] plumb `TimerMetric` through to Ruby- and Java- metrics APIs
- [x] migrate all internal timers
- [x] provide opt-in live-tracking implementation of `TimerMetric`
- [x] refactor metrics-related code from `QueueReadClient` up to `WorkerLoop` (or elsewhere)
- ~[ ] validate performance of live-tracking `TimerMetric`, and make it the default~ _[deferred: #14809]_

## How to test this PR locally

The metrics underneath `duration_in_millis` for filters, outputs, codecs, pipelines, and process have all been moved to the new timers API, so they should work with net zero change. Similarly `queue_push_duration_in_millis` for inputs, the pipeline, and the process have been moved.

I enjoy using a bursty pipeline:
~~~
input { java_generator { threads => 16 } }

filter {
  ruby {
    init => 'require "prime"'
    code => "
      if Prime.prime?(Time.now.min) && (Random.rand(100) <= 1)
        sleep(2.0 ** Random.rand(4.0))
      end
    "
  }
}

output { sink { } }
~~~

And observing the Node Stats API to see these metrics (and the flow metrics that rely on them) changing in real-time:

~~~
curl -XGET 'localhost:9600/_node/stats'
~~~

### CODE REVIEW

File: logstash-core/lib/logstash/instrument/metric_type.rb [hunk 1]
```diff
@@ -17,24 +17,23 @@
...
     # @param [String] The name of the class
     # @param [Array] Namespaces list
     # @param [String] The metric key
     # @raise [NameError] If the class is not found
     def self.create(type, namespaces, key)
-      METRIC_TYPE_LIST[type].new(namespaces, key)
+      case type
+      when :counter then return LogStash::Instrument::MetricType::Counter.new(namespaces, key)
+      when :gauge   then return LogStash::Instrument::MetricType::Gauge.new(namespaces, key)
+      when :uptime  then return org.logstash.instrument.metrics.UptimeMetric.new(key.to_s)
```
Comment (yaauie) [inline logstash-core/lib/logstash/instrument/metric_type.rb:33]: review note: `UptimeMetric` doesn't need a ruby-based delegator, so it was removed.

File: logstash-core/spec/logstash/util/wrapped_synchronous_queue_spec.rb [hunk 2]
```diff
@@ -80,12 +80,14 @@
             write_client.push_batch(batch)
 
             read_batch = read_client.read_batch.to_java
-            sleep(0.1) # simulate some work for the `duration_in_millis`
-            # TODO: this interaction should be cleaned in an upcoming PR,
-            # This is what the current pipeline does.
-            read_client.add_filtered_metrics(read_batch.filteredSize)
-            read_client.add_output_metrics(read_batch.filteredSize)
-            read_client.close_batch(read_batch)
+            read_client.execute_with_timers do
```
Comment (yaauie) [inline logstash-core/spec/logstash/util/wrapped_synchronous_queue_spec.rb:83]: this is pretty awful. We are effectively mimicking the specific chain of events that the `WorkerLoop` does, so this is liable to drift. Migrating these _out_ of the read client is deferred scope.

File: logstash-core/src/main/java/org/logstash/config/ir/compiler/JavaCodecDelegator.java [hunk 3]
```diff
@@ -71,52 +65,40 @@ public JavaCodecDelegator(final Context context, final Codec codec) {
...
 
     @Override
     public void encode(final Event event, final OutputStream out) throws IOException {
         encodeMetricIn.increment();
 
-        final long start = System.nanoTime();
-
-        codec.encode(event, out);
-
-        decodeMetricTime.increment(TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS));
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/config/ir/compiler/JavaCodecDelegator.java:119]: oh look, a bug. This is likely a part of why the encode metrics weren't getting tracked correctly.

File: logstash-core/src/main/java/org/logstash/instrument/metrics/MetricExt.java [hunk 4]
```diff
@@ -51,8 +53,9 @@ public final class MetricExt extends AbstractSimpleMetricExt {
     private static final RubySymbol DECREMENT = RubyUtil.RUBY.newSymbol("decrement");
 
     private static final RubySymbol GAUGE = RubyUtil.RUBY.newSymbol("gauge");
-
+    private static final RubySymbol TIMER = RubyUtil.RUBY.newSymbol("timer");
     private static final RubySymbol SET = RubyUtil.RUBY.newSymbol("set");
+    private static final RubySymbol GET = RubyUtil.RUBY.newSymbol("get");
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/instrument/metrics/MetricExt.java:58]: Oops. Not used (we need a java-`String`)

```suggestion
```

File: logstash-core/src/main/java/org/logstash/instrument/metrics/MetricExt.java [hunk 5]
```diff
@@ -140,30 +143,37 @@ protected IRubyObject getGauge(final ThreadContext context, final IRubyObject na
...
+
     @Override
     protected IRubyObject doReportTime(final ThreadContext context, final IRubyObject namespace,
         final IRubyObject key, final IRubyObject duration) {
         MetricExt.validateKey(context, null, key);
-        return collector.callMethod(
-            context, "push", new IRubyObject[]{namespace, key, COUNTER, INCREMENT, duration}
-        );
+
+        final TimerMetric timer = timer(context, namespace, key).toJava(TimerMetric.class);
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/instrument/metrics/MetricExt.java:161]: The dance between `timer` and `getTimer` is hard to follow, but mirrors the `gauge`/`getGauge` dance that is done throughout. I am certain this can be made more efficient, but equally certain that making sense of it all is beyond the scope I can handle in a single PR.

File: logstash-core/src/main/java/org/logstash/execution/WorkerLoop.java [hunk 6]
```diff
@@ -84,12 +84,15 @@ public void run() {
                 if (batch.filteredSize() > 0 || isFlush) {
                     consumedCounter.add(batch.filteredSize());
                     readClient.startMetrics(batch);
-                    final int outputCount = execution.compute(batch, isFlush, false);
-                    int filteredCount = batch.filteredSize();
-                    filteredCounter.add(filteredCount);
-                    readClient.addOutputMetrics(outputCount);
-                    readClient.addFilteredMetrics(filteredCount);
-                    readClient.closeBatch(batch);
+                    readClient.executeWithTimers(() -> {
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/execution/WorkerLoop.java:87]: This change is a little absurd, but currently the metrics are owned by the `QueueReadClient` so it is not net-worse. I would like to pull the metrics up to the worker loop (or perhaps elsewhere) in a second pass, but wanted this change to be distinctly visible.

File: logstash-core/src/main/java/org/logstash/execution/QueueReadClientBase.java [hunk 7]
```diff
@@ -196,7 +188,6 @@ public void rubyAddOutputMetrics(final IRubyObject size) {
     public void startMetrics(QueueBatch batch) {
         long threadId = Thread.currentThread().getId();
         inflightBatches.put(threadId, batch);
-        inflightClocks.put(threadId, System.nanoTime());
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/execution/QueueReadClientBase.java:199]: review note: we no longer put an inflight clock marker for the thread.

File: logstash-core/src/main/java/org/logstash/execution/QueueReadClientBase.java [hunk 8]
```diff
@@ -131,13 +130,6 @@ public IRubyObject rubyReadBatch(final ThreadContext context) throws Interrupted
     public void closeBatch(QueueBatch batch) throws IOException {
         batch.close();
         inflightBatches.remove(Thread.currentThread().getId());
-        Long startTime = inflightClocks.remove(Thread.currentThread().getId());
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/execution/QueueReadClientBase.java:134]: review note: when closing a batch, we no longer do manual time math, and are no longer responsible for removing a start time from the map because execution is wrapped in `executeWithTimers(Runnable)`.

**Comment by yaauie** [comment]
Jenkins test this again please

File: logstash-core/src/test/java/org/logstash/instrument/metrics/timer/TimerMetricTest.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+ * This {@code TimerMetricTest} is meant to be inherited by tests covering the implementations of {@link TimerMetric},
+ * and includes baseline tests for guaranteeing that the value of the metric includes the duration of
+ * execution before control is given back to the caller.
+ *
+ * <p>At an <em>interface</em>-level, we make no guarantees that tracked executions will be committed into the
+ * value <em>until</em> control is returned, which means that the value may or may not include "uncommitted" or
+ * currently-executing tracked timings. This means that our assertions can only apply when there are no concurrent
+ * executions happening.
+ */
+public abstract class TimerMetricTest {
```
Comment (mashhurs) [inline logstash-core/src/test/java/org/logstash/instrument/metrics/timer/TimerMetricTest.java:31]: Huge thanks for the clear test cases, especially blocked timed task tests are super helpful.

File: logstash-core/src/main/java/org/logstash/instrument/metrics/timer/TimerMetric.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    Long getValue();
+
+    @Override
+    default MetricType getType() {
+        return MetricType.TIMER_LONG;
+    }
+
+    static TimerMetric create(final String name) {
+        // INTERNAL-ONLY system property escape hatch, set with `metric.timers` config in logstash.yml
+        switch (System.getProperty("ls.metric.timers", "delayed")) {
```
Comment (mashhurs) [inline logstash-core/src/main/java/org/logstash/instrument/metrics/timer/TimerMetric.java:26]: What is the purpose of considering `ls.metric.timers`? For any input, we are returning `AfterCompletionTimerMetric`, removing confusing `switch/case` would be better if there is no any hidden logic.

Comment (yaauie) [reply logstash-core/src/main/java/org/logstash/instrument/metrics/timer/TimerMetric.java:26]: The meta-spec defines that we will provide a `metric.timers` config in the `logstash.yml` so that when we introduce live-tracking timers it can be selected.

This system property is set at startup with the value from the `metric.timer` setting, and is the injection point where we will add `live` timer routing. The PR as of when you reviewed it only had the one implementation so the switch statement is superfluous, but it is intended to make review of the additional implementation more straight-forward.

File: logstash-core/src/main/java/co/elastic/logstash/api/TimerMetric.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+ * or {@link NamespacedMetric#reportTime(String, long)}.
+ */
+public interface TimerMetric {
+    <T, E extends Throwable> T time(ExceptionalSupplier<T, E> exceptionalSupplier) throws E;
+
+    void reportUntrackedMillis(final long untrackedMillis);
+
+
+    default <E extends Throwable> void time(ExceptionalRunnable<E> exceptionalRunnable) throws E {
+        this.time(() -> { exceptionalRunnable.run(); return null; });
```
[the remark this answers is not in the document]
Comment (yaauie) [reply logstash-core/src/main/java/co/elastic/logstash/api/TimerMetric.java]: This one is tricky. If we don't return null, it maps to `TimerMetric#time(ExceptionalRunnable<E>)` which is self-referential. I've updated to include explicit type parameters, which makes the usage more clear.

**Comment by yaauie** [comment]
I moved the metric ownership out of the queue, but still need to detangle some of the tests that were improperly attempting to validate the worker loop's behaviour by duplicating all of its logic. It is possible that I'll need to revert [836ed21](https://github.com/elastic/logstash/pull/14748/commits/836ed21d94606bdc7858b5cf7bf7b502419e2cd1) until we have the capacity to validate that refactor.

**Comment by yaauie** [comment]
On hold until we merge your #14743. I can easily handle resolving the merge conflict on my end.

**Comment by yaauie** [comment]
Change of plans communicated with @mashhurs over separate channel:

I will merge this, then tackle the work of rebasing #14743 on top of it.
